### ISSUE TITLE
Mockable mutex wrapper with testing properties

### ISSUE THREAD

**Issue by richiware** [eProsima/Fast-RTPS#384]
Current architecture doesn't permit to test these new requirements. We cannot write a test to check an application's thread have never been blocked for an indeterminate time. To be able to do that, we need a infrastructure able to record the mutexes used by user's threads and to lock them at will.

Two libraries will be implemented to achieve this:

* mutex_testing_tool: contains the mutex testing infrastructure. Tests will link against this library and use its API to record mutexes and to lock them.
* mutex_testing_tool_preload: necessary to substitute the pthread functions `pthread_mutex_lock` and `pthread_mutex_timedlock`. Tests have to be run using the environment variable LD_PRELOAD to load this library.

**Mutex testing infrastructure API**
- *start_recording()* [Global function]: records in a global atomic variable the thread which wants to
record the mutex usages since that moment.
- *stop_recording()* [Global function]: sets in the global atomic variable the PID value 0.
  It indicates the recording stops.
- *get_num_mutexes()* [Global function]: returns the number of registered mutexes.
- *get_num_lock_type()* [Global function] : returns the number of mutexes that have used `pthread_mutex_lock()`.
- *get_num_timedlock_type()* [Global function] : returns the number of mutexes that have used `pthread_mutex_timedlock()`.
- *lock_mutex(size_t index)* [Global function]: locks the registered mutex.   After calling this function the test should check a call to verify it is not blocked for an indeterminate time.
- *unlock_mutex*(size_t index) [Global function]: unlocks the registered mutex.

Example:

```cpp
std::timed_mutex mutex_1;

tmutex_start_recording();

mutex_1.lock();
mutex_1.unlock();

tmutex_stop_recording();

tmutex_lock_mutex(1);

std::promise<int> promise_1;
std::future<int> future_1 = promise_1.get_future();
std::thread([&]{ mutex_1.try_lock_until(std::chrono::steady_clock::now() + std::chrono::milliseconds(100));
        promise_1.set_value_at_thread_exit(0); }).detach();

// This function will return because timeout of the std::timed_mutex.
std::cout << "Waiting..." << std::endl;
future_1.wait();

tmutex_unlock_mutex(1);
```

**Mutex testing preload library**

- *pthread_mutex_lock()*: if atomic variable has a valid PID then it will register the mutex and then lock the internal mutex.
- *pthread_mutex_timedlock()*: if atomic variable has a valid PID then it will register the mutex and then lock the internal mutex.

**Comment by richiware** [comment]
CC @dejanpan @anup-pem

**Comment by richiware** [comment]
First implementation #383

**Comment by richiware** [comment]
I made a quick search but I don't find a library that permits to lock mutexes inside a library. I want to test this implementation but using LD_PRELOAD to avoid the compiler option.

**Comment by richiware** [comment]
@dejanpan @anup-pem I've change the design and the implementation. Instead of a wrapper I directly overload `pthread_mutex_lock` using LD_PRELOAD.

**Comment by anup-pem** [comment]
@richiware I looked at your code changes at https://github.com/eProsima/Fast-RTPS/pull/383/files and it looks good. But, I am not very clear about the use case of this framework. 

I understand the recording part. After start and stop recording, we have a list of mutexes used by the application. After the recording, I am not sure about the usage part of that recorded mutexes list. I am sure I am missing something, which you might be able to help me out to answer my dumb questions.

My question is surrounding the explanation of the below design doc point and looking at the code in pull request #383 

> lock_mutex(size_t index) [Global function]: locks the registered mutex. After calling this function the test should check a call to verify it is not blocked for an indeterminate time.

Is this framework suppose to be used by an application developer to test just the locks defined and used by the developer or to test all the locks in the applications including from other libraries?

**Comment by richiware** [comment]
This framework is for using it in FastRTPS tests. We will create specific tests to check several functions are non-blocking.

Besides helping us to test now we achieve our goal of making several functions as non-blocking, the main purpose of these tests is detect we won't break new behavior in the future. Maybe including a new feature, a developer inserts a blocking `lock()` call and these tests should detect this.

These test will follow this steps:

1. Before calling the non-blocking call we start recording the mutexes this call will use.
2. Call the non-blocking call.
3. Stop recording.
4. Iterate locking each recorded mutex and calling the non-blocking call in another thread. The non-blocking call always should return and not keep blocked.

I hope this clarify

**Comment by anup-pem** [comment]
@richiware Thank you for your comments. Have a follow up question

# Scenario-1:

Let us take the example in this design article. I have incorporated my comments as I understand.

std::timed_mutex mutex_1;
tmutex_start_recording();
mutex_1.lock();
mutex_1.unlock();
tmutex_stop_recording();

******************** We have recorded the mutex used (i.e mutex_1).

tmutex_lock_mutex(1);   
**********************  Lock the recorded mutex (mutex_1)

std::promise<int> promise_1;
std::future<int> future_1 = promise_1.get_future();
std::thread([&]{ mutex_1.try_lock_until(std::chrono::steady_clock::now() + std::chrono::milliseconds(100));
        promise_1.set_value_at_thread_exit(0); }).detach();

********************* Since the mutex_1 is locked, the above thread will timeout after 100ms right? But, we don't know from the recorded info, how it was used during the recording phase. Was it used with a timeout or not. If my understanding is correct, what are we trying to test here? 

// This function will return because timeout of the std::timed_mutex.
std::cout << "Waiting..." << std::endl;
future_1.wait();

tmutex_unlock_mutex(1);

# Scenario-2:
Say developer-A creates a mutex and uses that mutex to lock. The developer might use a blocking lock API such as pthread_mutex_lock(), std::mutex or timed lock API's such as pthread_mutex_timed_lock() or std::timed_mutex() with some timeout. Now, developer-B or a application user, while testing the application, we first do the recording. It just stores the mutex used. It does not know how it was used in the application. 

I am still not clear how this framework helps in finding a blocking lock() in an application. I assume either the recording method should store more information about how the lock was used(it is challenging) or just use a profiling tool like perf or some other tool and check where in the application a blocking lock is called without timeouts.

Might be we can have a quick call to clarify my confusion.

**Comment by richiware** [comment]
@anup-pem The code you paste in Scenario-1 is based on my unit tests for this framework. These unit tests try to check the framework works for all STD mutex types. FastRTPS tests using this framework will have more checking.

But you're right. It is not complete yet. I didn't remember POSIX has the function `pthread_mutex_timedlock()`. I will have to overwrite this function also, and store more information as I did in first implementation.

The purpose is not only to check if a user's thread call a blocking lock(). Also to verify the worst case always spends the same time, in current implementation and in future adding new features. And for testing this, the framework permits to externally lock all mutex used in the FastRTPS API call.

**Comment by anup-pem** [comment]
@richiware 
> The code you paste in Scenario-1 is based on my unit tests for this framework. These unit tests try to check the framework works for all STD mutex types. FastRTPS tests using this framework will have more checking.

Got it. Thank you for the clarifying that

> The  purpose is not only to check if a user's thread call a blocking lock().  Also to verify the worst case always spends the same time, in current  implementation and in future adding new features. And for testing this,  the framework permits to externally lock all mutex used in the FastRTPS  API call.

I totally agree with your comment about what needs to be done. I still have few questions about the implementation part. Instead of me asking now where you are still in the middle of implementation, I will wait for you to implement it and use it in FastRTPS.  Once I see how you use it, my questions might get answered.
